Support generic stream/event on CUDA/HIP backend

guangyey opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125757

# Motivation
According to [#123611](https://github.com/pytorch/pytorch/pull/123611), we support generic stream/event on CUDA backend.

# Additional Context
new method/attribute on `torch.Event` for cuda
- torch.Event.event_id
- torch.Event.elapsed_time
- torch.Event.synchronize

new method on `c10::Event` on cuda backend
- c10.Event.event_id
- c10.Event.elapsed_time
- c10.Event.synchronize

guangyey committed:
Support generic stream/event on CUDA backend

[ghstack-poisoned]

guangyey committed:
Update on "[WIP] Support generic stream/event on CUDA backend"




[ghstack-poisoned]

albanD commented on c10/cuda/impl/CUDAGuardImpl.h:
```diff
@@ -205,11 +205,35 @@ struct CUDAGuardImpl final : public c10::impl::DeviceGuardImplInterface {
+        "Both events must be recorded before calculating elapsed time.");
+    cudaEvent_t cuda_event1 = static_cast<cudaEvent_t>(event1);
+    cudaEvent_t cuda_event2 = static_cast<cudaEvent_t>(event2);
+    float time_ms = 0;
```
Could you add the device guard like https://github.com/pytorch/pytorch/blob/1ecea513b648502e4a0765256d165db5d3b42f52/aten/src/ATen/cuda/CUDAEvent.h#L157 here? Looks important.

guangyey replied:
Updated. We also provide support on HIP backend.

albanD replied:
Why did you add get/set calls here and not `DeviceGuard guard(c10::kCUDA);` ?

guangyey replied:
In my opinion, `CUDAGuardImpl` is an implementation of `DeviceGuard`. `DeviceGuard` is high-level than `CUDAGuardImpl`. So it is unreasonable to use high-level code(`DeviceGuard`) in low-level code(`CUDAGuardImpl`).
If you prefer to use `DeviceGuard`, I can prepare a PR to refine it.

albanD reviewed (approved):
Thanks!

guangyey committed:
Update on "[WIP] Support generic stream/event on CUDA backend"




[ghstack-poisoned]

guangyey committed:
Update on "Support generic stream/event on CUDA backend"




[ghstack-poisoned]

guangyey committed:
Update on "Support generic stream/event on CUDA backend"




[ghstack-poisoned]

jgong5 reviewed (changes requested):
Test failing?

guangyey commented on torch/csrc/Event.cpp:
```diff
@@ -58,8 +58,11 @@ static PyObject* THPEvent_pynew(
       device->type(),
-      (enable_timing ? c10::EventFlag::PYTORCH_DEFAULT
-                     : c10::EventFlag::BACKEND_DEFAULT));
+      // See note [Flags defining the behavior of events]
```
bug fix

guangyey commented on aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h:
```diff
@@ -213,12 +213,35 @@ struct HIPGuardImplMasqueradingAsCUDA final : public c10::impl::DeviceGuardImplI
     hip_stream.synchronize();
   }
 
+  void synchronizeEvent(void* event) const override {
```
help to support hip backend.

guangyey committed:
Update on "Support generic stream/event on CUDA backend"


# Motivation
According to [#123611](https://github.com/pytorch/pytorch/pull/123611), we support generic stream/event on CUDA backend.

# Additional Context
new method/attribute on `torch.Event` for cuda
- torch.Event.event_id
- torch.Event.elapsed_time
- torch.Event.synchronize

new method on `c10::Event` on cuda backend
- c10.Event.event_id
- c10.Event.elapsed_time
- c10.Event.synchronize


[ghstack-poisoned]

guangyey commented:
@pytorchbot merge

guangyey commented:
@pytorchbot merge

guangyey committed:
Update on "Support generic stream/event on CUDA backend"


# Motivation
According to [#123611](https://github.com/pytorch/pytorch/pull/123611), we support generic stream/event on CUDA backend.

# Additional Context
new method/attribute on `torch.Event` for cuda
- torch.Event.event_id
- torch.Event.elapsed_time
- torch.Event.synchronize

new method on `c10::Event` on cuda backend
- c10.Event.event_id
- c10.Event.elapsed_time
- c10.Event.synchronize


[ghstack-poisoned]

guangyey committed:
Update on "Support generic stream/event on CUDA backend"


# Motivation
According to [#123611](https://github.com/pytorch/pytorch/pull/123611), we support generic stream/event on CUDA backend.

# Additional Context
new method/attribute on `torch.Event` for cuda
- torch.Event.event_id
- torch.Event.elapsed_time
- torch.Event.synchronize

new method on `c10::Event` on cuda backend
- c10.Event.event_id
- c10.Event.elapsed_time
- c10.Event.synchronize


[ghstack-poisoned]

guangyey committed:
Update on "Support generic stream/event on CUDA backend"


# Motivation
According to [#123611](https://github.com/pytorch/pytorch/pull/123611), we support generic stream/event on CUDA backend.

# Additional Context
new method/attribute on `torch.Event` for cuda
- torch.Event.event_id
- torch.Event.elapsed_time
- torch.Event.synchronize

new method on `c10::Event` on cuda backend
- c10.Event.event_id
- c10.Event.elapsed_time
- c10.Event.synchronize


[ghstack-poisoned]

guangyey commented:
@pytorchbot merge

albanD commented on aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h:
```diff
@@ -213,12 +213,36 @@ struct HIPGuardImplMasqueradingAsCUDA final : public c10::impl::DeviceGuardImplI
     HIPCachingAllocatorMasqueradingAsCUDA::recordStreamMasqueradingAsCUDA(data_ptr, hip_stream);
   }
+
+  double elapsedTime(void* event1, void* event2, const DeviceIndex device_index)
```
Can we please make sure the relevant submodule owner has a chance to review the change before we merge it in other submodules?

cc @jeffdaily does this timing code looks good?

jeffdaily replied:
Nothing jumps out as wrong.  I'll lean on CI to tell the truth :-).

guangyey replied:
> Can we please make sure the relevant submodule owner has a chance to review the change before we merge it in other submodules?
> 
> cc @jeffdaily does this timing code looks good?

Sorry about that...

guangyey replied:
I refer to `destroyEvent` code to guard it not to crate a new cuda context here. https://github.com/pytorch/pytorch/blob/b24ad7eab55eaf660893dddae949fc714e434338/c10/cuda/impl/CUDAGuardImpl.h#L106

guangyey committed:
Update on "Support generic stream/event on CUDA/HIP backend"


# Motivation
According to [#123611](https://github.com/pytorch/pytorch/pull/123611), we support generic stream/event on CUDA backend.

# Additional Context
new method/attribute on `torch.Event` for cuda
- torch.Event.event_id
- torch.Event.elapsed_time
- torch.Event.synchronize

new method on `c10::Event` on cuda backend
- c10.Event.event_id
- c10.Event.elapsed_time
- c10.Event.synchronize


[ghstack-poisoned]
